File: frontend/src/types/status.ts
import type { EntityConfig } from './domain';

export type RunState = 'setup' | 'printing' | 'proofing' | 'hold' | 'released';
export const ALL_RUN_STATE: readonly RunState[] = ['setup', 'printing', 'proofing', 'hold', 'released'];
export type DecisionType = 'release' | 'rework' | 'quarantine';
export const ALL_DECISION_TYPE: readonly DecisionType[] = ['release', 'rework', 'quarantine'];

export const ENTITY_CONFIGS: readonly EntityConfig[] = [
  { key: 'pressUnit', path: 'presses', label: '印刷设备', statuses: ['ready', 'setup', 'printing', 'maintenance'] as const },
  { key: 'printRun', path: 'runs', label: '印刷批次', statuses: ['setup', 'printing', 'proofing', 'hold', 'released'] as const },
  { key: 'colorProof', path: 'proofs', label: '色彩校样', statuses: ['captured', 'review', 'accepted', 'rejected'] as const },
  { key: 'releaseDecision', path: 'release', label: '放行决定', statuses: ['draft', 'release', 'rework', 'quarantine'] as const }
];

File: backend/internal/constants/status.go
package constants

// Shared status values are mirrored in frontend/src/types/status.ts. Keeping
// the lists explicit makes state-machine drift visible during code review.

type RunState string

const (
	RunStateSetup    RunState = "setup"
	RunStatePrinting RunState = "printing"
	RunStateProofing RunState = "proofing"
	RunStateHold     RunState = "hold"
	RunStateReleased RunState = "released"
)

var AllRunState = []string{"setup", "printing", "proofing", "hold", "released"}

type DecisionType string

const (
	DecisionTypeRelease    DecisionType = "release"
	DecisionTypeRework     DecisionType = "rework"
	DecisionTypeQuarantine DecisionType = "quarantine"
)

var AllDecisionType = []string{"release", "rework", "quarantine"}

type CalibrationStatus string

const (
	CalibrationStatusPending CalibrationStatus = "pending"
	CalibrationStatusPassed  CalibrationStatus = "passed"
	CalibrationStatusFailed  CalibrationStatus = "failed"
)

var AllCalibrationStatus = []string{"pending", "passed", "failed"}

// CalibrationTransitions models the closed loop: a pending request is resolved
// exactly once. A failed request may be reopened by a reviewer after the batch
// returns to proofing; passed requests are immutable evidence.
var CalibrationTransitions = map[string]map[string]bool{
	"pending": {"passed": true, "failed": true},
	"failed":  {"pending": true},
	"passed":  {},
}

var PressUnitTransitions = map[string]map[string]bool{
	"ready":       {"setup": true, "printing": true},
	"setup":       {"printing": true, "maintenance": true, "ready": true},
	"printing":    {"maintenance": true, "setup": true},
	"maintenance": {"printing": true},
}

var PrintRunTransitions = map[string]map[string]bool{
	"setup":    {"printing": true},
	"printing": {"proofing": true, "hold": true, "setup": true},
	"proofing": {"hold": true, "released": true, "printing": true},
	"hold":     {"proofing": true},
	"released": {"hold": true},
}

var ColorProofTransitions = map[string]map[string]bool{
	"captured": {"review": true},
	"review":   {"accepted": true, "rejected": true, "captured": true},
	"accepted": {"review": true},
	"rejected": {"review": true},
}

var ReleaseDecisionTransitions = map[string]map[string]bool{
	"draft":      {"release": true, "rework": true},
	"release":    {"rework": true, "quarantine": true, "draft": true},
	"rework":     {"quarantine": true, "release": true},
	"quarantine": {"rework": true},
}

func CanTransition(graph map[string]map[string]bool, from, to string) bool {
	targets, exists := graph[from]
	return exists && targets[to]
}
